roofit/roofit/src/RooLagrangianMorphFunc.cxx

```diff
@@ -0,0 +1,3544 @@
+    if (!first)
+      ss << sep;
+    ss << itr->GetName();
+    first = false;
+  }
+  return ss.str();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// this is a workaround for the missing implicit conversion from
+/// SuperFloat<>double
+
+template <class A, class B> inline void assignElement(A &a, const B &b) {
+  a = static_cast<A>(b);
+}
+///////////////////////////////////////////////////////////////////////////////
+// read a matrix from a stream
+
+template <class MatrixT>
+inline MatrixT readMatrixFromStreamT(std::istream &stream) {
+  std::vector<std::vector<SuperFloat>> matrix;
+  std::vector<SuperFloat> line;
+  while (!stream.eof()) {
+    if (stream.peek() == '\n') {
+      stream.get();
+      stream.peek();
+      continue;
+    }
+    SuperFloat val;
+    stream >> val;
+    line.push_back(val);
+    while (stream.peek() == ' ' || stream.peek() == '\t') {
+      stream.get();
+    }
+    if (stream.peek() == '\n') {
+      matrix.push_back(line);
+      line.clear();
+    }
+  }
+  MatrixT retval(matrix.size(), matrix.size());
+  for (size_t i = 0; i < matrix.size(); ++i) {
+    if (matrix[i].size() != matrix.size()) {
+      std::cerr << "matrix read from stream doesn't seem to be square!"
+                << std::endl;
+    }
+    for (size_t j = 0; j < matrix[i].size(); ++j) {
+      assignElement(retval(i, j), matrix[i][j]);
+    }
+  }
+  return retval;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// read a matrix from a text file
+
+template <class MatrixT> inline MatrixT readMatrixFromFileT(const char *fname) {
+  std::ifstream in(fname);
+  if (!in.good()) {
+    std::cerr << "unable to read file '" << fname << "'!" << std::endl;
+  }
+  MatrixT mat = readMatrixFromStreamT<MatrixT>(in);
+  in.close();
+  return mat;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// convert a TH1* param hist into the corresponding ParamSet object
+
+template <class T> inline std::map<const std::string, T> readValues(TH1 *h_pc) {
+  std::map<const std::string, T> point;
+  if (h_pc) {
+    // loop over all bins of the param_card histogram
+    for (int ibx = 1; ibx <= h_pc->GetNbinsX(); ++ibx) {
+      // read the value of one parameter
+      const std::string s_coup(h_pc->GetXaxis()->GetBinLabel(ibx));
+      double coup_val = h_pc->GetBinContent(ibx);
+      // add it to the map
+      if (!s_coup.empty()) {
+        point[s_coup] = T(coup_val);
+      }
+    }
+  }
+  return point;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// retrieve a param_hist from a certain subfolder 'name' of the file
+
+inline TH1F *getParamHist(TDirectory *file, const std::string &name,
+                          const std::string &objkey = "param_card",
+                          bool notFoundError = true) {
+  TFolder *f_tmp = dynamic_cast<TFolder *>(file->Get(name.c_str()));
```

Comment:
```suggestion
  auto f_tmp = file->Get<TFolder>(name.c_str());
```
Please use the typesafe `TDirectory::Get<>`, then you also save a `dynamic_cast`.